src/erlfmt.erl

```diff
@@ -272,11 +272,48 @@ format_file_range(FileName, StartLocation, EndLocation, Options) ->
 ) ->
     {ok, string(), [error_info()]}
     | {error, error_info()}.
-format_string_range(String, StartLocation, EndLocation, Options) ->
-    FileName = "nofile",
+format_string_range(Original, StartLocation, EndLocation, Options) ->
+    format_string_range("nofile", Original, StartLocation, EndLocation, Options).
+format_string_range(FileName, Original, StartLocation, EndLocation, Options) ->
     Pragma = proplists:get_value(pragma, Options, ignore),
-    {ok, Nodes, Warnings} = read_nodes_string(FileName, String, Pragma),
-    format_enclosing_range(FileName, StartLocation, EndLocation, Options, Nodes, Warnings).
+    {ok, Nodes, Warnings} = read_nodes_string(FileName, Original, Pragma),
+    {{StartLine, _}, {EndLine, _}, Result} = format_enclosing_range(
+        FileName,
+        StartLocation,
+        EndLocation,
+        Options,
+        Nodes,
+        Warnings
+    ),
+    case Result of
+        {ok, Formatted, Info} ->
+            Whole = inject_range(Original, StartLine, EndLine, Formatted),
+            {ok, Whole, Info};
+        Error ->
+            Error
+    end.
+
+% Reinject formatted extract into whole string.
+inject_range(Original, StartLine, EndLine, Formatted) ->
+    AsList = string:split(unicode:characters_to_binary(Original), "\n", all),
+    FormattedAsList0 = string:split(Formatted, "\n", all),
+    % Remove spurious empty line introduced by last \n separator.
+    FormattedAsList =
+        case lists:last(FormattedAsList0) of
+            <<>> ->
+                lists:droplast(FormattedAsList0);
+            _ ->
+                FormattedAsList0
+        end,
+    if
+        EndLine < StartLine ->
```

Comment:
Hmm good catch
I can't remember, do we maybe check this during argument parsing to give the user a nice error when this is not the case?

Reply:
can do that! Will also use the idiomatic construct suggested by @michalmuskala elsewhere. 
Would give:
`true = EndLine < StartLine`

Reply:
Well when it is not `true` then the error given to the user would be stack trace instead of a nicely formatted error, or am I misunderstanding?

Reply:
I forgot about that, but there is already an user-friendly check in `erlfmt_cli.erl`. It gives:
`Range: End must be greater or equal than Start`

Reply:
Great, then possibly we don't also need a check here?

Reply:
Asserts don't hurt and have documentation value! It's also useful if the function get plugged from another client than the CLI.